### PR TITLE
Usage count filter

### PR DESCRIPTION
**Pull request by FWDekker**
Using the visitor introduced in #234, this PR adds a filter that filters out patterns that do not have enough library usages.

### CODE REVIEW

**Comment by casperboone** [comment]
The code lgtm, however, what's the use of this filter? Why do we need it? :)

**Comment by FWDekker** [comment]
@casperboone Many of our patterns contain only one call to the library. This would result in EvoSuite generating tests that test the library rather than actual patterns. This filter removes such patterns by ensuring that all patterns interact with the library at least `n` times.

**Comment by casperboone** [comment]
@FWDekker I agree that we need to filter out patterns with just one usage, however right now the default is '3', which seems rather arbitrary.

**Comment by JSAbrahams** [comment]
Personally I would have the default at 2, which to me makes the most sense.

**Comment by FWDekker** [comment]
@casperboone Do you think 2 is less arbitrary? :P

**Comment by casperboone** [comment]
@FWDekker Yes, cause that's the lowest possible number for testing interactions between multiple library objects/methods/fields.

In general: I'm not sure whether we should do this kind of filtering at all. Even if there's just one call to a library method, it can still be very relevant to the library maintainer to know if his new implementation breaks on input cases that exist in user land.

**Comment by gandreadis** [comment]
@casperboone Fair point, although we have to see whether this throws more false than true positives in practice (looking at what EvoSuite gives us)

**Comment by FWDekker** [comment]
@casperboone But if a pattern has only one library usage then wouldn't that mean we're just testing the library, even if the input is based on actual user input? It doesn't sound like much of a pattern to me.

**Comment by FWDekker** [comment]
@casperboone @gandreadis I have fixed the issue by adding so-called "processors" after the filters in the JLUG. These filters replace the condition (for an if) or key (for a switch) with an unbound boolean (for an if) or unbound int (for a switch) that will be randomised by EvoSuite.

**Comment by FWDekker** [comment]
@casperboone @JSAbrahams How bout that?

File: modules/mining-pipeline/jimple-pattern-filter/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/InsufficientLibraryUsageFilterRule.kt [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ * A visitor that determines whether a given [Value] uses [libraryProject].
+ *
+ * @property libraryProject the project that should be used
+ */
+class LibraryUsageVisitor(private val libraryProject: JavaProject) : JimpleValueVisitor<Boolean>() {
+    /**
+     * Returns false.
+     *
+     * @param value a [Value]
+     * @return false
```
Comment (gandreadis) [inline modules/mining-pipeline/jimple-pattern-filter/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/InsufficientLibraryUsageFilterRule.kt]: ❓

**Comment by gandreadis** [comment]
Hmm, @FWDekker, just merged this, but now I'm wondering - have you also added this as option to the CLI? I don't see it being set somewhere, and the README.md isn't updated with the new option, either. Could you add that in another PR?

**Comment by FWDekker** [comment]
@gandreadis Whoops. Luckily I _did_ add the option in #309, so that should be OK. I forgot to update the README in that PR though.

**Pull request merged**